### ISSUE TITLE
Django and Flask local server gives Connection Reset Error

### ISSUE THREAD

**Issue by zags** [Microsoft/BashOnWindows#1707]
##  A brief description
I am having difficulty running a Django and Flask projects locally.  I keep getting a "The connection was reset" browser error rather than my site, though clicking refresh many times does occasionally succeed.  Given there is at least a 75% chance of a page failing to load and this happens on each page, this makes running sites locally unusable.

## Your Windows build number
14393.693

## Steps / All commands required to reproduce the error from a brand new installation 
Steps for Django:
Run project locally with `python manage.py runserver`
Go to `127.0.0.1:8000` in the browser

Steps for Flask:
Run project locally with `python app.py`
Go to  or `127.0.0.1:5000` in the browser

##  Strace of the failing command
When I get the "Connection was reset" error, both Flask and Django say the page was retrieved with a 200 status, but say the page was retrieved many times for one attempted load in the browser.  Here is the Django output for one browser request:

```
Performing system checks...

System check identified no issues (0 silenced).
February 15, 2017 - 21:00:15
Django version 1.10.5, using settings 'config.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
[15/Feb/2017 21:00:19] "GET / HTTP/1.1" 200 29
```

## Required packages and commands to install
This requires having python and either Django 1.10 or Flask 0.12 installed through pip
(`pip install Flask` or `pip install django`)

## Additional Information
Changing the port doesn't help.

Changing the configuration to have the sites listen on `0.0.0.0` doesn't help.

This occurs in both python 2 and python 3.

This occurs whether or not I am using a python virtual environment.

This occurs regardless of whether the files are on the Windows versus the Ubuntu side of the file system.

This occurs in both Firefox and Chrome.

Here are two unanswered Stackexchange posts about this issue:
http://superuser.com/q/1147168/605884
http://stackoverflow.com/q/39751907/2800876

I have no problem running rails sites locally through WSL.

**Comment by jakegt1** [comment]
I have been using flask recently with not many issues on build 14986 - what build are you on?

**Comment by zags** [comment]
as stated in the original post, I am on build 14393.693

**Comment by jakegt1** [comment]
Sorry, I missed that..

I'm going to take a educated guess and say it's an issue on build 14393, but personally I can't pinpoint the real reason for you.

There's a lot of socket/networking implementation between 14393 and 14986, so it might be something to do with that.

**Comment by stehufntdev** [comment]
Thanks for reporting the issue.  Marking this as fixed in insider builds based on jakeget1's comments.  Agreed there are a number of networking issues in the anniversary update (build 14393), so please try out an insider build or the creators update build when it is available.

**Comment by zags** [comment]
I've found a workaround for people experiencing this problem.  I can run sites locally by installing Apache and serving them with that.  Obviously this lacks many of the advantages of Django and Flask's ability to run sites locally, so while not ideal, this is at least functional.

**Comment by jakegt1** [comment]
out of interest, is that using mod_wsgi or something else?

**Comment by zags** [comment]
This is using mod_wsgi.  I installed apache2 and libappache2-mod-wsgi, symlinked my project into /var/www, and then used standard django apache configuration:

```
WSGIApplicationGroup %{GLOBAL}
LogLevel warn

<VirtualHost *:80>
    ServerName XXXXX
    WSGIDaemonProcess XXXXX user=XXXXX group=XXXXX threads=5 home=/var/www/projects/XXXXX python-path=/var/www/projects/XXXXX
    WSGIProcessGroup XXXXX
    Alias /static /var/www/projects/XXXXX/static
    WSGIScriptAlias / /var/www/projects/XXXXX/XXXXX/wsgi.py
    <Directory /var/www/projects/XXXXX>
        Require all granted
    </Directory>
</VirtualHost>
```

**Comment by tdifilippo** [comment]
Hi,

 I have run a flask-restful-api, and I get errors when trying to connect from Edge or Chrome to 127.0.0.1:5000. It never really responds to any endpoints, and if I run the same exact flask app in Windows, I get my restful endpoints to respond just fine. I have a question, is this supposed to be fixed in Windows DSL, or is it also the flavor of Ubuntu? I don't have he newest Windows. I am still stuck with the older DSL. Does anyone know how to change to the newer DSL (14965 and counting, I think). I changed my PC yesterday to Insiders, but I originally set the level to Release Preview, I quickly changed to Fast as soon as I discovered the new changes. I have updated my Ubuntu the painful way, looking through the blogs, I now have 16.04 Xenial running. How do I change the DSL? My goal is to change to a newer DSL and check whether this works, I would much rather use Ubuntu than Windows for my testing.

**Comment by tdifilippo** [comment]
PS, I am not using MOD_WSGI on Apache, just the flask version, if that makes a difference,

**Comment by aseering** [comment]
Hi @tdifilippo -- I'm not sure what you mean by "DSL"?  But if you want a newer build, you have to:

* Switch to the Insider Slow or Fast Lane
* Wait 24 to 48 hours for the change to take effect.  (And the computer can't be shut down for all of that time.)

Basically, as I understand it, your computer has to run some basic compatibility checks before it will let you install a newer build, and it's hardcoded to do so once per day.

If that doesn't work, you can download a .iso and install manually.  But the .iso's aren't as up-to-date, generally speaking.

**Comment by benhillis** [comment]
@aseering - I  believe .iso files are only released for Insider slow builds.

**Comment by aseering** [comment]
@benhillis -- for what it's worth, my observation has been that that is usually true but that sometimes they get out-of-sync -- either there's a special newer-than-Slow cut of the .iso, or the .iso has been delayed, etc.  (The OP referred specifically to the Fast ring; I figured ""aren't as up-to-date" wasn't wrong in that context, and was more concise than the above.)

I could certainly be misunderstanding, though.  You'd know better than I :-)

**Comment by tdifilippo** [comment]
Sorry, typo. I meant WSL, OK get it I have to wait like 24 to 48 hours, that's great. Thanks for the note about not shutting down. 
Sent from Yahoo Mail on Android 
 
  On Mon, Feb 27, 2017 at 10:55 AM, Adam Seering<notifications@github.com> wrote:   
Hi @tdifilippo -- I'm not sure what you mean by "DSL"? But if you want a newer build, you have to:
   
   - Switch to the Insider Slow or Fast Lane
   - Wait 24 to 48 hours for the change to take effect. (And the computer can't be shut down for all of that time.)

Basically, as I understand it, your computer has to run some basic compatibility checks before it will let you install a newer build, and it's hardcoded to do so once per day.

If that doesn't work, you can download a .iso and install manually. But the .iso's aren't as up-to-date, generally speaking.

**Comment by tdifilippo** [comment]
I guess I would have gotten the newer version if I had selected "fast track" right away. I'm
like to see whether the Windows 10, with Ubuntu bash will allow me stay off the VirtualBox Ubuntu. Would much rather use a quick bash command prompt than have to spin up a VM. I'm not sure, based on what I read about network support (ie no WiFi support) that I'll be able to get my browser in Windows to communicate with a rest-ful api running from flask on Ubuntu. 
Sent from Yahoo Mail on Android 
 
  On Mon, Feb 27, 2017 at 11:17 AM, Adam Seering<notifications@github.com> wrote:   
@benhillis -- for what it's worth, my observation has been that that is usually true but that sometimes they get out-of-sync -- either there's a special newer-than-Slow cut of the .iso, or the .iso has been delayed, etc. (The OP referred specifically to the Fast ring; I figured ""aren't as up-to-date" wasn't wrong in that context, and was more concise than the above.)

I could certainly be misunderstanding, though. You'd know better than I :-)
